### ISSUE TITLE
AI

### ISSUE THREAD

**Issue by yoonsch217** [yoonsch217/archive#5]

**Comment by yoonsch217** [comment]
# Machine Learning

https://wikidocs.net/32012

하이퍼파라미터(초매개변수) : 모델의 성능에 영향을 주는 사람이 값을 지정하는 변수.
매개변수 : 가중치와 편향. 학습을 하는 동안 값이 계속해서 변하는 수.

### 회귀 vs 분류
- 회귀 문제: Lineare Regression
연속적인 범위 내에서 예측값을 구한다.
- 분류 문제: Logistic Regression
  - binary classification
  - multi-class classification

### 지도 vs 비지도
- 지도 학습: Supervised Learning
레이블(Label)이라는 정답과 함께 학습하는 것
- 비지도 학습: Unsupervised Learning
- 자기지도 학습: Self-Supervised Learning, SSL
레이블이 없는 데이터가 주어지면, 모델이 학습을 위해서 스스로 데이터로부터 레이블을 만들어서 학습하는 경우를 자기지도 학습


### 벡터, 행렬, 텐서

벡터는 1차원 배열, 행렬은 2차원, 3차원부터는 tensor라고 부른다.
벡터를 1차원 텐서라고 부르기도 한다. 스칼라는 0차원 텐서이다.
n차원 벡터라고 한다면 size n짜리 리스트라는 뜻이다.


## Linear Regression

독립변수: x, 종속변수: y
- 단순 선형 회귀 분석(Simple Linear Regression Analysis)
y = wx + b
- 다중 선형 회귀 분석(Multiple Linear Regression Analysis)
y = w1x1 + w2x2 + ... + b

머신러닝은 여기서 최적의 w, b를 찾는 것이 목표이다.


### Cost Function

실제값과 예측값에 대한 오차에 대한 식을 목적 함수(Objective function) 또는 비용 함수(Cost function) 또는 손실 함수(Loss function)라고 합니다.
함수의 값을 최소화하거나, 최대화하거나 하는 목적을 가진 함수를 목적 함수(Objective function)라고 합니다. 그리고 값을 최소화하려고 하면 이를 비용 함수(Cost function) 또는 손실 함수(Loss function)라고 합니다.

- MSE
- MAE
- ...


### Optimizer

선형 회귀를 포함한 수많은 머신 러닝, 딥 러닝의 학습은 결국 비용 함수를 최소화하는 매개 변수인 w와 b을 찾기 위한 작업을 수행합니다. 이때 사용되는 알고리즘을 옵티마이저(Optimizer) 또는 최적화 알고리즘이라고 부릅니다.
이 옵티마이저를 통해 적절한 w와 b를 찾아내는 과정을 머신 러닝에서 훈련(training) 또는 학습(learning)이라고 부릅니다.

- Gradient Descent
(기본 모델)w에 대한 cost function 그래프를 그려보면 포물선 모양일 것이다. 포물선의 꼭짓점에서의 w값이 cost function을 최소화시키는 w 값이다.
`next w = previous w - alpha * (cost function을 w로 편미분한 뒤 previous w 넣은 값, 즉 previous w일 때의 기울기)` 으로 업데이트한다. alpha: learning rate
- ...


## Logistic Regression

선형으로는 classification 문제를 풀기 어렵다. true, false를 위해서는 output 값을 0에서 1로 제한하는 게 편한데 선형의 경우는 output 값이 무한대로 갈 수 있기 때문이다.
하나의 예시로 0에서 1 사이의 값을 갖는 시그모이드 함수를 사용할 수 있다. 
그렇게 되면 1에 가까울 수록 true, 0에 가까울 수록 false일 확률이 높다고 볼 수 있다.


logistic regression의 경우 cost function으로 MSE를 사용하지 않는다. 꼭짓점이 여러 개인데 최소가 아닌 꼭짓점에 머무르게 되는 local minimum에 빠질 가능성이 있기 때문이다.



### Softmax Regression

이진 분류가 아니라 세 가지의 선택지가 있다면 어떨까.
각각에 대해 이진분류를 통해 true일 확률을 구하고 그 중 가장 높은 걸 고르는 방법도 있다.
하지만 softmax regression을 사용하면 각각이 될 확률을 구해줘서 총합이 1인 결과를 보여준다.

**Comment by yoonsch217** [comment]
# Deep Learning

Perceptron: 다수의 입력으로부터 하나의 결과를 내보내는 알고리즘
Step function: 여러 입력값이 각각의 가중치 weight이 곱해져서 출력된다. 이 출력값이 threshold(theta)를 넘으면 출력으로 1을 출력한다.

즉, `if sigma(wi*xi) >= theta, y = 1` 이다.
임계치 theta를 bias로 표현한다면, `if sigma(wi*xi) + b >= 0, y = 1`로 표현할 수 있다.

이렇게 뉴런에서 출력값을 변경시키는 함수를 활성화 함수(Activation Function)라고 한다.
퍼셉트론의 활성화 함수는 계단 함수이지만 여기서 활성화 함수를 시그모이드 함수로 변경하면 방금 배운 퍼셉트론은 곧 이진 분류를 수행하는 로지스틱 회귀와 동일하다.

### Single-Layer Perceptron

input layer와 output layer만 갖는다.
- input layer: 값을 보내는 단계
- output layer: 값을 받아서 출력하는 단계

AND, OR, NAND는 가능하지만 XOR은 안 된다.
각각에 대해 좌표에 그려보면 총 네 개의 점이 나오고 각각의 점은 true, false를 갖는다. 이 때, SLP로 분류를 한다는 건, 직선으로 true는 true끼리, false는 false끼리 묶어야한다는 것이다. XOR의 경우는 대각선으로 되어 있어서 직선으로 분류가 불가능하다는 것을 알 수 있다.
이런 경우는 다층 퍼셉트론을 사용하여 여러 개의 선으로 분류하는 효과를 얻을 수 있다.

### MultiLayer Perceptron, MLP

XOR 게이트는 AND, NAND, OR 게이트를 조합하여 만들 수 있다. 즉, 입력층과 출력층 사이에 hidden layer를 더 쌓는 것이다.

hidden layer가 두 개 이상인 신경망을 DNN(Deep Neural Network, 심층 신경망) 이라고 한다.
이런 신경망에서 가중치를 스스로 찾도록 자동화한 과정이 deep learning이다.


### Artificial Neural Network

- 피드 포워드 신경망(Feed-Forward Neural Network, FFNN): 오직 출력층 방향으로 연산이 전개된다. RNN은 FFNN이 아니다.
- 전결합층(Fully-connected layer, FC, Dense layer): 어떤 층의 모든 뉴런이 이전 층의 모든 뉴런과 연결돼 있는 층을 전결합층(Fully-connected layer) 또는 완전연결층이라고 한다.
- 활성화 함수(Activation Function)
  - non linear 해야한다. step function 또한 비선형이긴 하다. 활성화 함수로 선형 wx를 쓰게 되면 hidden layer를 추가할 때 `w*w*..*w*x` 로 되는데 이는 곧 `k*x`와 다를 게 없다. 




### 순전파(Foward Propagation)

입력 차원 3, 출력 차원 2라고 하면 x1, x2, x3를 받아서 y1, y2를 출력한다는 것이다.
이 경우 parameter의 수는 8개가 된다.
입력들이 y1으로 가는 화살표에서 w 3개, b 1개가 있고 마찬가지로 y2로 갈 때도 4 개가 있다.

![image](https://user-images.githubusercontent.com/41099071/191496452-84cba276-71b1-459b-a19b-a26182f4ae00.png)

이건 입력 샘플 하나만 처리할 때의 모습이다.
여러 샘플을 동시에 병렬로도 처리할 수가 있다. 이렇게 인공 신경망이 다수의 샘플을 동시에 처리하는 것을 우리는 '배치 연산'이라고 부른다.

![image](https://user-images.githubusercontent.com/41099071/191496640-028d6014-d5b4-4b6f-bbc0-8dec1db54c3d.png)

순전파를 진행하고 예측값을 구하고나서 이 다음에 인공 신경망이 해야할 일은 예측값과 실제값으로부터 오차를 계산하고, 오차로부터 가중치와 편향을 업데이트하는 일이다. 이때 인공 신경망은 순전파와는 반대 방향으로 연산을 진행하며 가중치를 업데이트하는데, 이 과정을 역전파(BackPropagation)라고 한다.


### Loss Function


- MSE(Mean Squared Error, MSE): 연속형 변수를 예측할 때 사용
- 이진 크로스 엔트로피(Binary Cross-Entropy): 분류
- 카테고리칼 크로스 엔트로피(Categorical Cross-Entropy): multi-class classification

### Epoch, Batch Size, Iteration(Step)

- Epoch: 인공 신경망에서 전체 데이터에 대해서 순전파와 역전파가 끝난 상태이다.
- Batch Size: 가중치 등의 매개 변수의 값을 조정하기 위해 사용하는 데이터의 양이다. 
- Iteration: 한 번의 epoch을 끝내기 위해 필요한 배치의 수이다. 데이터가 2천개이고 배치 크기가 200이라면 배치의 수는 10개이다. 이 배치의 수를 이터레이션이라고도 한다.

배치 크기에 따른 Gradient Descent의 특징
- 배치 경사 하강법(Batch Gradient Descent)
`batch_size = len(X_train)` 전체 입력 크기를 batch size로 설정한다.
배치 경사 하강법은 옵티마이저 중 하나로 오차(loss)를 구할 때 전체 데이터를 고려한다. 한 번의 에포크에 모든 매개변수 업데이트를 단 한 번 수행한다.
전체 데이터를 고려해서 학습하므로 한 번의 매개 변수 업데이트에 시간이 오래 걸리며, 메모리를 크게 요구한다는 단점이 있다.
- 배치 크기가 1인 확률적 경사 하강법(Stochastic Gradient Descent, SGD)
랜덤으로 선택한 하나의 데이터에 대해서만 계산하는 방법으로, 더 빠르게 계산할 수 있다.
- 미니 배치 경사 하강법(Mini-Batch Gradient Descent)
직접 batch size를 정해주는 방법으로서, 전체 데이터를 계산하는 것보다 빠르며, SGD보다 안정적이다.
2의 거듭제곱의 수로 하는 것이 일반적이다.




### Optimizer

- 모멘텀(Momentum)
경사 하강법에서 계산된 접선의 기울기에 한 시점 전의 접선의 기울기값을 일정한 비율만큼 반영한다.  local minimum을 탈출할 힘을 줄 수도 있다. `tf.keras.optimizers.SGD(lr=0.01, momentum=0.9)`
- 아다그라드(Adagrad)
각 매개변수에 서로 다른 학습률을 적용한다. 변화가 많은 매개변수는 학습률이 작게 설정되고 변화가 적은 매개변수는 학습률을 높게 설정된다. 
- 알엠에스프롭(RMSprop)
아다그라드는 학습을 계속 진행한 경우에는, 나중에 가서는 학습률이 지나치게 떨어진다는 단점이 있는데 이를 다른 수식으로 대체하여 이러한 단점을 개선했다.
- 아담(Adam)
아담은 알엠에스프롭과 모멘텀 두 가지를 합친 듯한 방법으로, 방향과 학습률 두 가지를 모두 잡기 위한 방법이다.
`tf.keras.optimizers.Adam(lr=0.001, beta_1=0.9, beta_2=0.999, epsilon=None, decay=0.0, amsgrad=False)`
즉, Adam도 GD 방법을 응용한 하나의 optimizer네?




### BackPropagation


### Overfitting 막는 방법

- 데이터 수 늘리기
- 모델의 복잡도 줄이기(hidden layer, params 수)
- Regularization 적용하기: 매개변수가 많은 복잡한 모델이 overfitting될 가능성이 높다. 모델을 간단히 하는 방법으로 regularization이 있다.
- Dropout: 신경말 이부 사용 안 한다.



### Vanishing Gradient & Exploding

Gradient Descent할 때 미분 값이 너무 작으면 backPropagation하면서 점점 소실되는데 이게 gradient vanishing이다.
반대로 발산하여 Gradient exploding이 일어나기도 한다.

이에 따라서 Activation function를 사용하는 방법이 달라져야한다.

- 시그모이드 함수는 미분값이 최대 0.25이고 양 끝은 0에 가깝다. 따라서 시그모이드 함수를 활성화 함수로하는 인공 신경망의 층을 쌓는다면, 가중치와 편향을 업데이트 하는 과정인 back propagation 과정에서 0에 가까운 값이 누적해서 곱해지게 되면서, 앞단에는 기울기(미분값)가 잘 전달되지 않게 된다. 따라서 시그모이드는 hidden layer에서의 사용이 지양되고 출력 층에서 사용된다.
- Tanh 함수의 경우는 미분 최대값이 1이므로 소실이 상대적으로는 적은 편이다.
- ReLU는 hidden layer로 가장 인기 있다. `f(x) = max(0, x)` 연산 속도도 빠른데 입력이 음수일 땐 미분값이 0이 되어 회생이 어렵다.(dying ReLU)
- Leaky ReLU는 dying ReLU를 보완하기 위해 만들어졌다. `f(x) = max(ax, x), a는 작은 값`
- Softamx function도 시그모이드와 같이 출력층에서 자주 사용된다. 딥 러닝으로 이진 분류를 할 때는 출력층에 배운 로지스틱 회귀를 사용하고, 딥 러닝으로 다중 클래스 분류 문제를 풀 때는 출력층에 소프트맥스 회귀를 사용한다.


이를 방지하기 위한 여러 방법들도 있다.

- 그래디언트 클리핑(Gradient Clipping): 기울기 폭주를 막기 위해 임계값을 넘지 않도록 값을 자른다. 
`Adam = optimizers.Adam(lr=0.0001, clipnorm=1.)`
- 가중치 초기화(Weight initialization): 가중치 초기화만 적절히 해줘도 기울기 소실 문제과 같은 문제를 완화시킬 수 있다.
  - 세이비어 초기화(Xavier Initialization)
  - He 초기화(He initialization)
- 배치 정규화(Batch Normalization): 인공 신경망의 각 층에 들어가는 입력을 평균과 분산으로 정규화하여 학습을 효율적으로 만든다.
RNN은 각 시점(time step)마다 다른 통계치를 가지기 때문에 RNN에 배치 정규화를 적용하는 것을 어렵게 만든다. Layer Normalization을 사용해야한다.
  - 내부 공변량 변화(Internal Covariate Shift)
  - 배치 정규화(Batch Normalization)
- 층 정규화(Layer Normalization)

**Comment by yoonsch217** [comment]
# 순환 신경망(Recurrent Neural Network, RNN)

RNN은 은닉층의 노드에서 활성화 함수를 통해 나온 결과값을 출력층 방향으로도 보내면서, 다시 은닉층 노드의 다음 계산의 입력으로 보내는 특징을 갖고있다.
은닉층에서 활성화 함수를 통해 결과를 내보내는 역할을 하는 노드를 셀(cell)이라고 한다. 이 셀은 이전의 값을 기억하려고 하는 일종의 메모리 역할을 수행하므로 이를 메모리 셀 또는 RNN 셀이라고 표현한다.

메모리 셀은 각각의 시점(time step)에서 바로 이전 시점에서의 은닉층의 메모리 셀에서 나온 값을 자신의 입력으로 사용하는 재귀적 활동을 한다.
메모리 셀이 출력층 방향 또는 다음 시점인 t+1의 자신에게 보내는 값을 은닉 상태(hidden state) 라고 한다.
t 시점의 메모리 셀은 t-1 시점의 메모리 셀이 보낸 은닉 상태값을 t 시점의 은닉 상태 계산을 위한 입력값으로 사용한다.

![image](https://user-images.githubusercontent.com/41099071/191505914-3e675ae5-cb76-46f5-93d0-969112e49698.png)

여기서 t는 다른 샘플을 말하는 게 아니라 같은 샘플에서의 다른 시간의 값을 말하는 것 같다.

![image](https://user-images.githubusercontent.com/41099071/191507169-d6396e31-70ce-4e73-a515-3576dcb76d80.png)

단어 시퀀스에 대해서 하나의 출력을 하는 다 대 일(many-to-one) 구조의 모델은 입력 문서가 긍정적인지 부정적인지를 판별하는 감성 분류(sentiment classification), 또는 메일이 정상 메일인지 스팸 메일인지 판별하는 스팸 메일 분류(spam detection) 등에 사용할 수 있다.
이게 지금 내가 원하는 종류의 구조가 된다.

![image](https://user-images.githubusercontent.com/41099071/191511075-c1fc5691-e4e3-4582-bfb2-674e6d12a3a2.png)


- 은닉층: `h(t) = tanh(Wx * x(t) + Wh * h(t-1) + b)`
- 출력층: `y(t) = f(Wy * h(t) + b)`, f는 비선형 활성화 함수

은닉상태 h가 다음 은닉상태와 현재 입력 모두에 영향을 준다.


### 장단기 메모리(Long Short-Term Memory, LSTM)

바닐라 RNN(가장 단순한 RNN)의 한계를 극복하기 위해 나온 RNN의 변형이다.
바닐라 RNN의 시점(time step)이 길어질 수록 앞의 정보가 뒤로 충분히 전달되지 못하는 현상이 발생한다.
앞 부분에 중요한 데이터가 있는 경우 모델의 정확도에 문제가 생길 수 있다. 이를 장기 의존성 문제(the problem of Long-Term Dependencies)라고 한다.

LSTM은 은닉층의 메모리 셀에 입력 게이트, 망각 게이트, 출력 게이트를 추가하여 불필요한 기억을 지우고, 기억해야할 것들을 정한다.
은닉 상태(hidden state)를 계산하는 식이 전통적인 RNN보다 조금 더 복잡해졌으며 셀 상태(cell state)라는 값을 추가했다.

![image](https://user-images.githubusercontent.com/41099071/191511130-06e89786-f49d-4937-8aa7-bae3c4b8082d.png)
